lib/ui/painting/fragment_program.cc

```diff
@@ -19,6 +22,70 @@ namespace flutter {
 
 IMPLEMENT_WRAPPERTYPEINFO(ui, FragmentProgram);
 
+void FragmentProgram::initFromAsset(std::string asset_name) {
+  std::shared_ptr<AssetManager> asset_manager = UIDartState::Current()
+                                                    ->platform_configuration()
+                                                    ->client()
+                                                    ->GetAssetManager();
+  std::unique_ptr<fml::Mapping> data = asset_manager->GetAsMapping(asset_name);
+  if (data == nullptr) {
+    Dart_ThrowException(tonic::ToDart("Asset '" + asset_name + "' not found"));
+    return;
+  }
+
+  auto runtime_stage = impeller::RuntimeStage(std::move(data));
+  {
+    auto code_mapping = runtime_stage.GetCodeMapping();
+    auto code_size = code_mapping->GetSize();
+    const char* sksl =
+        reinterpret_cast<const char*>(code_mapping->GetMapping());
+    // SkString makes a copy.
+    SkRuntimeEffect::Result result =
+        SkRuntimeEffect::MakeForShader(SkString(sksl, code_size));
+    if (result.effect == nullptr) {
+      Dart_ThrowException(tonic::ToDart(std::string("Invalid SkSL:\n") + sksl +
+                                        std::string("\nSkSL Error:\n") +
+                                        result.errorText.c_str()));
+      return;
+    }
+    runtime_effect_ = result.effect;
+  }
+
+  int sampled_image_count = 0;
+  size_t other_uniforms_bytes = 0;
+  for (const auto& uniform_description : runtime_stage.GetUniforms()) {
+    if (uniform_description.type ==
+        impeller::RuntimeUniformType::kSampledImage) {
+      sampled_image_count++;
+    } else {
+      size_t size = uniform_description.dimensions.rows *
```

Comment:
@chinmaygarde Does this make sense? This is attempting to compute how many bytes are needed for all of the uniforms by adding up `(rows * cols * bit_width) / 8`. Does that make sense?

Reply:
Makes sense to me.